r/R/singlecell_utils.R

```diff
@@ -101,9 +101,13 @@ pseudobulk_matrix <- function(mat, cell_groups, method = "sum", threads = 1L) {
     }
   }
   assert_is(threads, "integer")
-  # if multiple methods are provided, only need to pass in the top method as it will also calculate the less complex stats
-  iter <- iterate_matrix(parallel_split(mat, threads, threads*4))
-  res <- pseudobulk_matrix_cpp(iter, cell_groups = as.integer(cell_groups) - 1, method = method, transpose = mat@transpose)
+
+  it <- mat %>%
+    convert_matrix_type("double") %>%
+    parallel_split(threads, threads*4) %>%
```

Comment:
Think this is reasonable to use the piping style instead.  

From what I understand, this forces all matrices no matter what type to be converted to doubles and sent to `pseudobulk_matrix<double>`. While this completely works, I question whether it makes sense to require usage of `pseudobulk_matrix<double>` as we have explicit instantation for `float` and `uint32_t`.  

I wonder if it makes sense to just properly handle this at the `matrix_utils.cpp` file level on `pseudobulk_matrix_cpp()`. Rather than forcing everything to work on matrices of type `double`, could we do something like how `SEXP iterate_matrix_col_select_<type>_cpp()` handles this, by using a different function depending on type, and doing a `do.call()` to perform a string method search for the correct function?  Either that, or passing a `type` param in to `pseudobulk_matrix_cpp()`, to determine which `pseudobulk_matrix` cpp type to call.

Reply:
Adding in proper type handling would be a great option. Right now several parts of BPCells do conversions to doubles for math, which is not terrible for efficiency when most of the calculations return doubles anyhow, but there's definitely room for improvement.

Properly utilizing the non-double implementations would be an interesting bit of follow-up work. With the plans for the C++/R/python refactor it might largely solve itself, but it's also not too hard to fix up in the current code base.